### DISCUSSION TITLE
Function to exit the program without a panic?

### DISCUSSION THREAD

**Discussion by tryoxiss** [gleam-lang/gleam#3559] [Questions & support] [answered]
Is there a way to exit a gleam program or task without getting an error? You can `panic as`, but that gives an error. `todo as` gives both an error and warning. This is far from ideal if the exit was intentional and not due to an error (e.g., a graceful shutdown) If not, is there a reason, or could it be added?

If such a function is added, an assosiated type should be added, being simply a type with no variants - making it impossible to construct, clearly showing that the function will not return (simple demo implementation below)

```
import gleam/io

pub type Never {}

pub fn main() -> Never {
  io.println("Do Stuff")
  exit()
}

pub fn exit() -> Never {
  panic as "exited" // replace with the proper way to exit a program if there is one
}
```

**Comment by Enderchief** [comment]
What would be a use case for this?

Maybe FFI could work instead. Although I've heard early exits aren't the best for the Erlang VM.
```gleam
// main.gleam
import gleam/io

pub fn main() {
  io.println("Hello!")
  exit(0)
  io.println("this never prints")
}

@external(erlang, "exit_ffi", "do_exit")
@external(javascript, "./exit_ffi.mjs", "do_exit")
pub fn exit(exit_code: Int) -> Nil
```

```erlang
% exit_ffi.erl
-module(exit_ffi).
-export([do_exit/1]).

do_exit(Code) ->
  halt(Code).
```
```js
// exit_ffi.mjs
export function do_exit(code) {
  if (globalThis.process) process.exit(code);
  else if (globalThis.Deno) Deno.exit(code);
  else throw Error; // if in the browser
}
```

**Reply by lpil** [reply]
I'd probably use init:stop rather than erlang:halt as it shuts down application trees

**Reply by tryoxiss** [reply]
> What would be a use case for this?
> 
> Maybe FFI could work instead. Although I've heard early exits aren't the best for the Erlang VM.
> 
> ```gleam
> // main.gleam
> import gleam/io
> 
> pub fn main() {
>   io.println("Hello!")
>   exit()
>   io.println("this never prints")
> }
> 
> @external(erlang, "exit_ffi", "do_exit")
> @external(javascript, "./exit_ffi.mjs", "do_exit")
> pub fn exit(exit_code: Int) -> Nil
> ```
> 
> ```erlang
> % exit_ffi.erl
> -module(exit_ffi).
> -export([do_exit/0]).
> 
> do_exit(Code) ->
>   halt(Code).
> ```
> 
> ```js
> // exit_ffi.mjs
> export function do_exit(code) {
>   if (globalThis.process) process.exit(code);
>   else if (globalThis.Deno) Deno.exit(code);
>   else throw Error; // if in the browser
> }
> ```

The case I had in mind was a server which is expected to be running until an exit is called for, at which point it should finish outstanding I/O and close itself. You mentioned early exists are not good for the BEAM VM, is there a better way to do this instead then?

**Reply by lpil** [reply, marked as answer]
> The case I had in mind was a server which is expected to be running until an exit is called for, at which point it should finish outstanding I/O and close itself.

Then you want init.stop, not erlang.halt

**Reply by tryoxiss** [reply]
> > The case I had in mind was a server which is expected to be running until an exit is called for, at which point it should finish outstanding I/O and close itself.
> 
> Then you want init.stop, not erlang.halt

Thank you!
